### PR TITLE
ASoC: SOF: handle paused streams during system suspend

### PR DESCRIPTION
**Pull request by ranj063**
During system suspend, paused streams do not get suspended.
Therefore, we need to explicitly free these PCMs in the DSP
and free the associated DAPM widgets so that they can be set
up again during resume.

Fixes https://github.com/thesofproject/linux/issues/3116

### CODE REVIEW

File: sound/soc/sof/sof-audio.c [hunk 1]
```diff
@@ -718,6 +718,59 @@ int sof_set_up_pipelines(struct snd_sof_dev *sdev, bool verify)
...
+				}
+			}
+		}
+
+	/*
+	 * free the DAI widget associated with the paused streams. This is equivalent to the
+	 * handling of suspend trigger for the BE DAI for running streams.
+	 */
+	list_for_each_entry(swidget, &sdev->widget_list, list)
+		if (WIDGET_IS_DAI(swidget->id) && swidget->use_count > 0) {
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c]: I am a bit nervous about this wildcard for all use_count > 0.
That might prevent us from catching some bugs?
Should this be done only for the use_count == 1?

File: sound/soc/sof/sof-audio.c [hunk 2]
```diff
@@ -718,6 +718,59 @@ int sof_set_up_pipelines(struct snd_sof_dev *sdev, bool verify)
...
+				if (ret < 0)
+					return ret;
+
+				ret = sof_widget_list_free(sdev, spcm, dir);
+				if (ret < 0) {
+					dev_err(sdev->dev, "failed to free widgets during suspend\n");
+					return ret;
+				}
+			}
+		}
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c:754]: How is this going to work if we have a mixer and host-dai split pipelines? Would we also free the paused BE?

File: sound/soc/sof/pm.c [hunk 3]
```diff
@@ -191,7 +191,7 @@ static int sof_suspend(struct device *dev, bool runtime_suspend)
 	if (sdev->fw_state != SOF_FW_BOOT_COMPLETE)
 		goto suspend;
 
-	/* set restore_stream for all streams during system suspend */
+	/* prepare for streams to be resumed properly upon resume */
```
Comment (plbossart) [inline sound/soc/sof/pm.c:194]: separate patch, there's no functionality change here.

File: sound/soc/sof/sof-audio.c [hunk 4]
```diff
@@ -752,6 +805,18 @@ int sof_tear_down_pipelines(struct snd_sof_dev *sdev, bool verify)
...
+	 * and unset the prepare flag so that they can be set up again during resume.
+	 */
+	if (!verify) {
+		ret = sof_tear_down_paused_pipelines(sdev);
+		if (ret < 0) {
+			dev_err(sdev->dev, "failed to tear down paused pipelines\n");
+			return ret;
+		}
+	}
+
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c:815]: sorry, I see nothing in this patch that tests if the status is PAUSED? Is this intentional, and if yes maybe a comment would be desired to guide reviewers and future code changes.

File: sound/soc/sof/sof-audio.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:754]: yes but only the last PCM that drops the ref_count for the BE will free the DAI.

File: sound/soc/sof/sof-audio.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: yes, I can change it to == 1.

File: sound/soc/sof/sof-audio.c (hunk 4 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:815]: maybe not clear still but I did mention why it is not necessary to check the state on line 733. let me reword it a bit

File: sound/soc/sof/sof-audio.c [hunk 5]
```diff
@@ -718,6 +718,59 @@ int sof_set_up_pipelines(struct snd_sof_dev *sdev, bool verify)
...
+	 */
+	list_for_each_entry(swidget, &sdev->widget_list, list)
+		if (WIDGET_IS_DAI(swidget->id) && swidget->use_count == 1) {
+			ret = sof_widget_free(sdev, swidget);
+			if (ret < 0) {
+				dev_err(sdev->dev, "failed to free DAI widget %s during suspend\n",
+					swidget->widget->name);
+				return ret;
+			}
+		}
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c:765]: I am still unclear on how the tests above apply to the PAUSED case only. 

the 'free the paused PCMS and associated DAPM widgets' is at best done with an indirect set of criteria?

Comment (ranj063) [reply sound/soc/sof/sof-audio.c:765]: @plbossart Im not sure Im following your question. But tear_dwn_pipelines() will only be called during system suspend and more specifically when the DSP should be in D3 during system suspend. In this case, all running streams will be suspended. The only spcm's that will have a widget list that is non-empty will be the paused ones. I could use the state check for paused streams but I think its unnecessary. But if you think it makes following the logic easier, I will make the change

Comment (plbossart) [reply sound/soc/sof/sof-audio.c:765]: I would feel more comfortable if we test the state and flag an error if the status is not paused. That said, if this is racy then it'd be a bad idea and we should only add a comment along the lines of what you wrote "The only spcm's that will have a widget list that is non-empty will be the paused ones"

Comment (ranj063) [reply sound/soc/sof/sof-audio.c:765]: i dont think it is a racy. let me modify it

Comment (ranj063) [reply sound/soc/sof/sof-audio.c:765]: but the reason I did not do that is because we are manually forcing the actions of the FE suspend and BE suspend on any streams that were left behind. I cant check the state for the for the first loop to walk through the spcms but the second one to free the DAI widgets doesnt have that option. 

If it makes it better, I could change the name of the function to not call it out tear_down_paused_pipelines() to tear_down_all_pipelines() or something similar and add in the comment that it will take care of paused pipelines.

Comment (plbossart) [reply sound/soc/sof/sof-audio.c:765]: not able to follow the comment "but the reason I did not do that is because we are manually forcing the actions of the FE suspend and BE suspend on any streams that were left behind. I cant check the state for the for the first loop to walk through the spcms but the second one to free the DAI widgets doesnt have that option."

in the paused case, there is no action of the FE and BE suspend.

Comment (ranj063) [reply sound/soc/sof/sof-audio.c:765]: right.  Ideally if the paused streams were also suspended, we'd have no problems. but now we're forced to do that ourselves to clean up all the paused streams. What I was trying to saw was that a pCM has a paused state but a DAI doesnt. SO adding the check only for the spcm loop and not for the bottom loop to free the DAI would look odd.

Comment (ranj063) [reply sound/soc/sof/sof-audio.c:765]: I've modifed the patch now to rename the function @plbossart . Let me know if this reads better

Comment (plbossart) [reply sound/soc/sof/sof-audio.c:765]: looks good and I tested it, works fine in the simple pipeline case. My device needs to recharge so I couldn't test the case where we have two PCMs (0 and 31) for the deep-buffer case that are paused when suspending. In this case would the widget count of 1 be accurate for the 'dai pipeline'. Or would this be 2?

Comment (ranj063) [reply sound/soc/sof/sof-audio.c:765]: no, I think it would still be 1. both PCMs would increment the DAI widget's in FE hw_params but the BE hw_params would only be called once.

Comment (plbossart) [reply sound/soc/sof/sof-audio.c:765]: that's the contentious part if you see earlier comments. In theory it's acceptable to do hw_params multiple times, but for cases with several FEs connected to the same BE it'd throw the refcounts in the weeds.

Comment (ranj063) [reply sound/soc/sof/sof-audio.c:765]: let me double check this part

Comment (plbossart) [reply sound/soc/sof/sof-audio.c:765]: I removed the filter with hw_params and this seems to work fine. The requirement is really that the CPU dais can tolerate multiple calls and internally keep the refcounts accurate. We might need to update this for HDaudio/DMIC/SSP but that's small potatoes compared to changing the soc-pcm.c code.

Comment (ranj063) [reply sound/soc/sof/sof-audio.c:765]: @plbossart I checked the deep buffer pipeline with both PCM's paused and suspend/resume worked just fine. The ref count is 1 when both PCMs are paused. This i because when the first BE hw-Params happens, we set the configured flag for the DAI and do nothing when the second one comes along.

Comment (plbossart) [reply sound/soc/sof/sof-audio.c:765]: can you provide a pointer on this? this doesn't seem to be aligned with what I see in soc-pcm.c

Comment (plbossart) [reply sound/soc/sof/sof-audio.c:765]: see e.g. soc_pcm_hw_params(), the comment says "This function can also be called multiple times" and internally it calls hw_params for all CPU dais.

Comment (ranj063) [reply sound/soc/sof/sof-audio.c:765]: I was wrong. I was getting confused with the hw_free where we check if the configred flag is set before freeing the widget.

But when the second BE hw_params comes along, it checks if the state of the all connected FE's are not active in https://github.com/thesofproject/linux/blob/c4bbda2ec4366f821eebdf2c2785dbdde854b03f/sound/soc/soc-pcm.c#L2890

I dont see a second BE hw_params in SOF and I'm thinking it is because of this.

Comment (plbossart) [reply sound/soc/sof/sof-audio.c:765]: There is still a corner case where the application could do multiple hw_params on the two devices, *before* starting one. 
I think it'd need dedicated hacks to test.

But even if that happens, I think the only issue is at the CPU dai level. if we do an internal hw_free to keep the refcounts consistent, regardless of what happens in the ALSA/ASoC core, we should be in good shape. It's done for SoundWire, I am not sure about the other DAIs.

Comment (ranj063) [reply sound/soc/sof/sof-audio.c:765]: > There is still a corner case where the application could do multiple hw_params on the two devices, _before_ starting one.
> I think it'd need dedicated hacks to test.

agree. Let me add a fix for this.

Comment (ranj063) [reply sound/soc/sof/sof-audio.c:765]: just added a patch for this @plbossart

File: sound/soc/sof/intel/hda.c [hunk 6]
```diff
@@ -58,6 +58,13 @@ int hda_ctrl_dai_widget_setup(struct snd_soc_dapm_widget *w)
 		return -EINVAL;
```
Comment (kv2019i) [inline sound/soc/sof/intel/hda.c:58]: The wording "unusual" is a bit unusual. :) I mean it is allowed and it does happen, so this is fixing a bug. But yeah, I get the point.
